work with public fairseq, fix bugs and add new features

znculee opened the pull request:
This pull request updates the constrained decoding code to be compatible with the latest [fairseq](https://github.com/pytorch/fairseq) (8185722), fix the tail stuttering bug, add the special evaluation for e2e dataset, update the post-filtering to pre-filtering, eliminate the inconsistency between TreeAcc and constrained decoding, add configurable order constraint for ds_join and add some bash scripts for the easier getting started.

Commit:
:recycle: be compatible with the public fairseq

- finalize hypotheses if all beams are not satisfied the constraints
- fix the conflict between finalizing failure cases and applying eos constraints
- fix inconsistant decoding with various batch_size
- add scripts for using out of the box

jinfengr commented:
Hi @znculee, thanks so much for integrating with fairseq! Can you help verify the results of constrained decoding of your implementation on the three datasets? Thanks a lot!

znculee committed:
:white_check_mark: update evaluation scripts

for e2e evaluation:
- group the hypotheses and references with the same original MRs
- select the best hypothesis in the group of the same original MRs by sentence BLEU
- evaluate the best hypotheses with the multi-references

for all evaluations:
- calculate BLEU with the unpreprocessed reference

znculee committed:
:boom: prefilter the non-terminals before beam search

features:
- checking whether each non-terminal is possible before beam search step
- constraint_penalty as ndarray
- no failure case

problems:
- checking all non-terminals is too slow

znculee committed:
:zap: accelerate non-terminal prefiltering by nomination

updates:
- nominate possible non-terminals instead of checking all ones
- regard step >= max_len as failure cases

problems:
- performance is very sensitive to beam size
- BLEU may not surpuss the postfilter approach

znculee committed:
:bug: fix tail stuttering

- guarantee to generate eos once it is permited
- allow closing basket only when all descents are covered or can be aggregated

an example of tail stuttering:

----------
| SOURCE |
----------
[0__DG_YES__  ]
[1__DS_JUSTIFY__
  [2__DG_RECOMMEND__
    [3__ARG_ATTIRE__ jacket ]
    [4__ARG_DATE_TIME_RANGE__
      [5__ARG_START_TIME__ 09:00 PM ]
      [6__ARG_COLLOQUIAL__ midnight ]
    ]
  ]
  [7__DG_INFORM__
    [8__ARG_DATE_TIME_RANGE__
      [9__ARG_START_TIME__ 09:00 PM ]
      [10__ARG_COLLOQUIAL__ midnight ]
    ]
    [11__ARG_TEMP_SUMMARY__ mid 30s ]
  ]
]

--------------
| HYPOTHESIS |
--------------
[0__DG_YES__ Yes ] ,
[1__DS_JUSTIFY__
  [2__DG_RECOMMEND__ you'll need a
    [3__ARG_ATTIRE__ jacket ]
    [4__ARG_DATE_TIME_RANGE__
      [6__ARG_COLLOQUIAL__ midnight ]
    ]
  ] because
  [7__DG_INFORM__ temperatures will be in the
    [11__ARG_TEMP_SUMMARY__ mid <number> ]
  ]
] !! STUTTERING !!

znculee committed:
:white_check_mark: add evaluation excluding failure cases

znculee committed:
:bug: eliminate inconsistency between TreeAcc and ConstrDec

- reimplement the TreeAcc by sharing the TreeConstraints class for
  constrained decoding
- add a script to print tree with proper indents
- update README

znculee committed:
:sparkles: add configurable order constraint

order constraint:
- force to generate the children of DS_JOIN in order
- activate this option by --order-constr in both
  fairseq-generate or compute_tree_acc

znculee committed:
:sparkles: use --dataset-impl raw

- use raw data for training and generating
- add an evaluation for constrained decoding by replacing failures

znculee committed:
:art: format the scripts

- replace `...` by $(...)
- add replacing failures to default evaluations

znculee committed:
:art: simplify awk program
